WireAPI/Sources/WireAPI/Public/APIs/BackendInfoAPI.swift

```diff
@@ -0,0 +1,27 @@
+//
+// Wire
+// Copyright (C) 2024 Wire Swiss GmbH
+//
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with this program. If not, see http://www.gnu.org/licenses/.
+//
+
+import Foundation
+
+public protocol BackendInfoAPI {
+
+    /// Fetch the info of the local backend.
+
+    func getBackendInfo() async throws -> BackendInfo
```

Comment:
opinion/idea: This declaration looks like its implementation should actually be put into a use case.
I wonder if we could reduce the level of abstraction.

Reply:
If a type conforming to this protocol lives in the API package/module, what would a use case implementation using this look like?
It would probably almost be a one-liner, calling this function.

I wonder if we could use the types of the iOS SDK directly instead of creating levels ob abstractions.

Reply:
What about a different approach. Every use of our backend API in the end follows this pattern:
- create an instance of `URLRequest`
- pass it to an `URLSession` instance
- handle the response (status code and bytes)

Why not putting that into a protocol, like:

```
protocol BackendAPI {
    associatedtype APIModel

    func createRequest(baseURL: URL) -> URLRequest
    func parse(data: Data, response: URLResponse) async throws -> APIModel
}
```

If dealing with the `URLSession` is repetitive, we can of course put another convenience implementation around this.

Reply:
Simple APIs like the one you implemented here could be put into a simple enum.

```
enum BackendInfoAPI: BackendAPI {

    case v0, v1, v2, v3, v4, v5

    func createRequest(baseURL: URL) -> URLRequest {
        /* create the instance */
    }

    func parse(data: Data, response: URLResponse) throws -> BackendInfo {
        /* we know
         - how to parse the bytes into our model
         - which status code to expect and how to translate them into error types
         */
    }

    struct BackendInfo {
        /* API model contents */
    }
}
```

Reply:
APIs that changed with the versions could be put into structs.

```
enum MoreComplexAPI {
    struct MoreComplexAPIModel { /* ... */ }
}

// MARK: V0

extension MoreComplexAPI {

    struct MoreComplexAPIV0: BackendAPI {

        func createRequest(baseURL: URL) -> URLRequest { /* ... */ }
        func parse(data: Data, response: URLResponse) throws -> MoreComplexAPIModel { /* ... */ }
    }
}

// MARK: V1

extension MoreComplexAPI {

    struct MoreComplexAPIV1: BackendAPI {

        func createRequest(baseURL: URL) -> URLRequest { /* ... */ }
        func parse(data: Data, response: URLResponse) throws -> MoreComplexAPIModel { /* ... */ }
    }
}
```

Something I'm just coming up with, maybe I'm missing a something important.

Reply:
> I wonder if we could use the types of the iOS SDK directly instead of creating levels of abstractions.

@caldrian this would bypass our network stack, so we'd need to reimplement quite a few things to keep existing behavior.

Reply:
> What about a different approach. Every use of our backend API in the end follows this pattern:
> - create an instance of URLRequest
> - pass it to an URLSession instance
> - handle the response (status code and bytes)

This is what @markusfassbender is also suggesting (right?)

Reply:
@johnxnguyen how will this WireApi module  fit in with the rest of the app? Is the plan that the `WireTransport` will go away eventually and that functionality  will live inside the WireAPI module?

Reply:
I haven't thought that much about it yet so but that is certainly an interesting idea. What do you think?

Reply:
See https://github.com/wireapp/wire-ios/pull/1297#discussion_r1591996064